dynamicio/mixins/with_s3.py

```diff
@@ -158,39 +159,66 @@ def _write_to_s3_path_prefix(self, df: pd.DataFrame):
+        downloads.
+
+        The `dynamic_file_path` supports pattern matching and variables (e.g., `part_{runner_id}.parquet`). This
+        approach  enables the downloading of files that specifically match the given pattern, optimizing I/O for
+        scenarios involving large datasets or multiple runners.
+
+        The method dynamically invokes appropriate file reading functions based on the `file_type` specified in the
+        configuration, supporting formats such as 'parquet', 'csv', 'hdf', and 'json'.
+
+        The function also includes an option to minimize disk space usage (`no_disk_space`). This is particularly
+        useful when needing to read a subset of columns from large files, thereby reducing the overall disk footprint.
 
-        To actually read the file, a method is dynamically invoked by name, using
-        "_read_{file_type}_path_prefix".
+        Parameters:
+        - None
 
         Returns:
-            DataFrame
+        - DataFrame: A pandas DataFrame concatenated from the read files.
+
+        Raises:
+        - ValueError: If the `file_type` specified in the configuration is not supported.
+
+        Configuration Keys:
+        - `bucket` (str): Name of the S3 bucket.
+        - `path_prefix` (str, optional): Static path prefix in the S3 bucket for file downloads.
+        - `dynamic_file_path` (str, optional): Dynamic file path with pattern matching for selective downloading of files.
+        - `file_type` (str): Type of the file to read ('parquet', 'csv', 'hdf', 'json').
+
+        Notes:
+        - Only one of `path_prefix` and `dynamic_file_path` can be provided.
+        - The function intelligently handles the download of files by synchronizing only those that match the specified
+        pattern in `dynamic_file_path`.
+        - e.g. a `runner_id` or any other variable used in `dynamic_file_path` for pattern matching should be specified
+        in the `options` of the configuration.
         """
         s3_config = self.sources_config.s3
         file_type = s3_config.file_type
         if file_type not in {"parquet", "csv", "hdf", "json"}:
             raise ValueError(f"File type not supported: {file_type}")
 
         bucket = s3_config.bucket
+        dynamic_file_path = s3_config.dynamic_file_path
         path_prefix = s3_config.path_prefix
-        full_path_prefix = utils.resolve_template(f"s3://{bucket}/{path_prefix}", self.options)
+
+        if dynamic_file_path:
+            full_path = utils.resolve_template(f"s3://{bucket}/{dynamic_file_path}", self.options)
+        else:
+            full_path = utils.resolve_template(f"s3://{bucket}/{path_prefix}", self.options)
 
         # The `no_disk_space` option should be used only when reading a subset of columns from S3
-        if self.options.pop("no_disk_space", False):
+        if self.options.pop("no_disk_space", False) and path_prefix:
```

Comment:
for future reference: mutating the dict (via `pop()`) in an `if` condition is pretty evil